compiler/src/diagnostics/output.re

```diff
@@ -43,9 +55,42 @@ let exn_to_lsp_error = (exn: exn): option(lsp_error) => {
 // kept lenses as an empty array so older LSPs don't crash
 
 let result_to_json =
-    (~errors: list(lsp_error), ~values: list(Lenses.lens_t)) => {
-  let result: lsp_result = {errors, values, lenses: []};
+    (
+      ~errors: list(lsp_error),
+      ~warnings: list(lsp_warning),
+      ~values: list(Lenses.lens_t),
+    ) => {
+  let result: lsp_result = {errors, warnings, values, lenses: []};
   Yojson.Basic.pretty_to_string(
     Yojson.Safe.to_basic(lsp_result_to_yojson(result)),
   );
 };
+
+let convert_warnings = (warnings_this_run, topLevelFileName) => {
+  List.fold_left(
+    (acc, w) => {
+      let (loc: Grain_utils.Warnings.loc, warn: Grain_utils.Warnings.t) = w;
+      let (file, line, startchar) =
+        Grain_parsing.Location.get_pos_info(loc.loc_start);
+      let (_, endline, endchar) =
+        Grain_parsing.Location.get_pos_info(loc.loc_end);
+
+      if (file == topLevelFileName) {
```

Comment:
Perhaps we should isolate the warnings arrays between files? Maybe something like this:
```reason
let with_preserve_warnings: (unit => 'a) => 'a;

// ...

// (in module_resolution.re:383)
with_preserve_unit^(~loc, chosen_unit_name, srcpath, () =>
  Grain_utils.Warnings.with_preserve_warnings(() => Grain_utils.Config.preserve_config(() =>
    compile_module_dependency^(srcpath, outpath)
  ))
);
```
This way, we shouldn't need to do this filtering step.

Reply:
Thanks for the discussion on this, looks a good way to handle it

Reply:
PR updated with the suggested method that works well.